### ISSUE TITLE
Process data on CudaSurface

### ISSUE THREAD

**Issue by AlphaStream** [halide/Halide#4600]
I have an image on a cuda surface and I'm trying to use Halide to process it.

I couldn't find a way to invoke Halide directly on the surface, so I implemented the following flow:
1. Copy image from surface to a cuda array
2. Create a Halide Buffer around that array
3. Run Halide
4. Copy from the output array back to the surface.

However, it appears that the output array is all 0's. Is this a correct approach at all?

```
// The surface we want to process.
cudaSurfaceObject_t surface(/* ... */);

void * input_array;
cudaMalloc(&input_array, sizeof(uint8_t) * image_width * image_height);

void * output_array;
cudaMalloc(&output_array, sizeof(uint8_t) * image_width * image_height);

CopySurfaceToArray(surface, (uint8_t*)input_array);
cudaDeviceSynchronize();

Halide::Runtime::Buffer<uint8_t> input_buffer(
  (uint8_t*)input_array, image_width, image_height);
Halide::Runtime::Buffer<uint8_t> output_buffer(
  (uint8_t*)output_array, image_width, image_height);

halide_blur(input_buffer, output_buffer);

CopyArrayToSurface((uint8_t*)output_array, surface);
cudaDeviceSynchronize();

cudaFree(input_array);
cudaFree(output_array);
```

If I replace `input_array` with `output_array` in the first call to `CopySurfaceToArray`, then I'll get the image unprocessed (as opposed to black screen), which suggests two things:
1. `CopyArrayToSurface` and `CopySurfaceToArray` work as expected
2. Halide just does not write anything to the `output_array`.

**Comment by abadams** [comment]
Those Buffer constructors expect CPU-side pointers (i.e. something returned by malloc). To make a Halide Buffer that wraps existing GPU memory created with cudaMalloc, you want something like:

```
Halide::Runtime::Buffer<uint8_t> input_buffer(nullptr, image_width, image_height);
input_buffer.device_wrap_native(halide_cuda_device_interface(), (uint64_t)input_array);
```

Finally, halide_blur launches work asynchronously, so if things still aren't working it's possible you need a cudaDeviceSynchronize() before the CopyArrayToSurface call (or call output_buffer.device_sync()).

**Comment by AlphaStream** [comment]
Thanks @abadams ! `device_wrap_native` did the trick.

**Comment by AlphaStream** [comment]
It turns out that copying from `cudaSurfaceObject__` to a bare CUDA array, and then back, takes even slightly more time than the actual Halide code (2.9ms vs 2.6ms).

Is it at all possible to make Halide take a `cudaSurfaceObject_t` directly as input, and also as output? 

I have found that CUDA surfaces wrap arrays created using `cudaMallocArray` (vs `cudaMalloc`). The code in the original post wraps Halide Buffers around `cudaMalloc`-arrays. However, when I tried to do the same with `cudaMallocArray`-array, the program crashed.

I guess it boils down to whether Halide supports generating code using Texture Memory read/write api.
